### DISCUSSION TITLE
Sharing custom function implementations with community

### DISCUSSION THREAD

**Discussion by 3bst0r** [hibernate/hibernate-orm#4889] [Design Proposals]
Hi,

Many databases support more functions/operators than the common denominator that hibernate, HQL, etc. support.
E.g., Postgis has '<->' to support KNN queries. 

As I understand the recommended way to make such functions/operators usable in HQL is to implement a custom SQLFunction and register it with a custom dialect, which I've tried and it works fine.

How would we share such extensions with the community? I am thinking about contributing a custom dialect and the custom functions for postgis, but I wanted to hear your opinions on how this could be meaningful for more people to use and less people writing the same "boilerplate" function implementations.

**Comment by sebersole** [comment]
Support for custom operators in HQL is not something I am going to do.  As it seems you tried already, you'd define a custom function which renders itself using the operator syntax and use that in your query.

We could implement support for such a thing through Criteria via custom nodes, although for consistency it is honestly probably best to simply do the same in Criteria as well.  Implementing a custom Criteria Expression node will require a bit of internals knowledge (e.g. you'd need it to extend `SelfRenderingExpression`).

As far as "sharing".. I assume you mean contributing?  I don't think these make sense in `hibernate-core` proper.  Maybe in `hibernate-spatial`, but I'll defer to @maesenka for that.  In terms of it not being accepted upstream, you could also host a project which adds such capabilities.  

I am also making the assumption that you are referring to 6.0, since these Discussions are limited to 6.0.  In which case, all of these things are much more pluggable.  For functions, it is as simple as leveraging Java's `ServiceLoader` mechanism to provide a `FunctionContributor`.

**Comment by 3bst0r** [comment]
Hi @sebersole  and thanks for the reply. 

I agree that custom functions don't have a place in hibernate-core or even hibernate-spatial, even though I am curious about @maesenka thoughts. So I propose another project, that I'd be willing to initiate, that acts as a "plugin" to hibernate for various database-specific function extensions.

The implementation side of adding operators or functions would be rather trivial I assume (whether it's a 6.0 `FunctionContributor` or a 5.x `SQLFunction`), it's the publishing and compatibility that I currently don't have a grip on. How would this project ensure and communicate compatibility between Hibernate versions? Are there any other examples for such "plugins" to Hibernate that I could take inspiration from?

As a side question, is your comment about 6.0 only referring to HQL, and not to Criteria? Or is Criteria also more pluggable in 6.0?

**Reply by beikov** [reply]
I am sure @maesenka would be happy to hear concrete proposals for HQL functions that you want to supported and I think that hibernate-spatial is the right place for adding support for PostGIS functions/operators we don't yet support. Let's start by listing them and the translation to SQL that you envision.

>How would this project ensure and communicate compatibility between Hibernate versions? Are there any other examples for such "plugins" to Hibernate that I could take inspiration from?

If you want to support multiple versions, you'll probably have to create dedicated modules like the hibernate-types project. IMO for 6.0 you can contribute this to hibernate-spatial directly though.

>As a side question, is your comment about 6.0 only referring to HQL, and not to Criteria? Or is Criteria also more pluggable in 6.0?

Not sure what you mean by pluggable exactly, but for HQL we don't want to support custom operators and instead model this through function syntax. As for Criteria, we ran out of time for 6.0 regarding the special pluggable support, but we might deliver this in a fix version maybe. Also see: https://github.com/hibernate/hibernate-orm/discussions/3819

**Reply by sebersole** [reply]
> adding operators or functions would be rather trivial

For HQL that is completely not true.  Adding operators is beyond trivial and not going to happen generally speaking

**Reply by 3bst0r** [reply]
> > adding operators or functions would be rather trivial
> 
> For HQL that is completely not true. Adding operators is beyond trivial and not going to happen generally speaking

@sebersole I thought I made it clear that I don't propose HQL to include more operators. I simply want to use some db specific operators without having to resort to native queries, and I want other people to have that too.

This is what I meant when I said „trivial implementation“:

```java
@Override
public String render(Type type, List arguments, SessionFactoryImplementor sessionFactoryImplementor) throws QueryException {
        return arguments.get(0) + "<->" + arguments.get(1);
}
```

Don't get me wrong but I don't see what's „completely not true“ about that.

**Reply by sebersole** [reply]
Good luck

**Reply by beikov** [reply]
I think there was just some misunderstanding here. @3bst0r wants to propose new SQM functions for hibernate-spatial, which I think is fine.

**Reply by 3bst0r** [reply]
@beikov 
> I think that hibernate-spatial is the right place for adding support for PostGIS functions/operators we don't yet support. Let's start by listing them and the translation to SQL that you envision.

Do you suggest adding functions to hibernate-spatial to be used in HQL that are only valid for one db, i.e. PostgreSQL? The concrete operators I was missing that led me to start this discussion are the distance operators from here https://postgis.net/docs/reference.html#Operators


> If you want to support multiple versions, you'll probably have to create dedicated modules like the hibernate-types project. IMO for 6.0 you can contribute this to hibernate-spatial directly though.

Thanks for the hint about hibernate-types, I'll be sure to study that in detail if we decide to go down that road of spawning another project instead of contributing something to hibernate-spatial.

>  for HQL we don't want to support custom operators and instead model this through function syntax

That's completely fine and I didn't want to suggest anything else.

**Reply by beikov** [reply]
We have standard distance and st_distance functions already, so all we need to do is change the rendering to the operator for PostGIS which should be quite easy. This is why I am asking for a list of functions you are envisioning. Chances are good that we have standard functions already that maybe just need a little tweaking

**Reply by maesenka** [reply]
I agree with @beikov that Hibernate Spatial is the right place for the Postgis functions. In H6 additional functions can easily be registered in `PostgisSqmFunctionDescriptors`.  

@3bst0r  which functions do you want added to the dialect?

**Reply by maesenka** [reply]
I didn't immediately see your answer, @3bst0r. There are some semantic differences between these operators and the st_distance function.  So they should be added as separate functions. We will need good names for them. Maybe you can suggest these?

**Reply by 3bst0r** [reply]
@maesenka here are my suggestions. `pgop`would stand for "postgis operator". The names afterwards take inspiration from how the postgis reference named their urls, like here https://postgis.net/docs/geometry_distance_cpa.html

`<->`: `pgop_distance_2d`
`|=|`: `pgop_distance_cpa`
`<#>`: `pgop_distance_box_2d`
`<<->>`: `pgop_distance_centroid_nd`
`<<#>>`: `pgop_distance_box_nd`

What do you think?

**Reply by maesenka** [reply]
I'm not a great fan w.r.t the 'pgop' prefix, to be honest.  I don't think it should matter that functions translate to operators. Would there be a difference between (e.g.) `pgop_distance_cpa` and `pg_distance_cpa`? Won't the prefix 'pg' (for postgis-specific) suffice?

**Reply by 3bst0r** [reply]
Sounds good to me!

**Reply by 3bst0r** [reply]
@maesenka is there already an issue that tracks these additions?

**Reply by beikov** [reply]
Maybe it's a stupid question, but isn't the PostGIS operator approach based on a sphere which other database might also support? For example MariaDB: https://mariadb.com/kb/en/st_distance_sphere/

So let's maybe name the functions accordingly and try to use the "standard naming" for these operators.

* `st_distance_sphere` renders to `<->`
* `st_distance_cpa` renders to `|=|`
* etc.

If we can maybe emulate these functions somehow through the standard set of functions, that would make this even desirable, but I don't know enough about the function to understand if that is possible.

**Reply by 3bst0r** [reply]
Not a stupid question. It boils down to the question if `<->` can be substituted for https://postgis.net/docs/ST_DistanceSphere.html in any case. I assume not, because that would mean `<->` is merely syntactic sugar, and the docs don't say that. It would also mean that postgis could have implemented KNN search using `ST_DistanceSphere` in the `ORDER BY`, but it doesn't. I  will try to find out more about the exact differences when I have time.

**Reply by beikov** [reply]
AFAIK PostgreSQL has a problem with indexes and functions which is why the operator is needed, but I'm pretty sure that all operators have a corresponding function name.

**Reply by maesenka** [reply]
@3bst0r I've created issue [HHH-15160](https://hibernate.atlassian.net/browse/HHH-15160) for this.

I don't mind to expand the set of standard functions beyond what we currently have, and `st_distancesphere`  is indeed a good candidate. 

However, I've been looking at the documentation and code of the '<->' operator. The operator is syntactic sugar, but the function invoked is different when used on Geometries, resp. Geographies. On Geometries the function `geometry_distance_centroid` is invoked, but on Geographies it's `geography_distance_knn`.  

Now we don't (yet) have a distinction between Geometry/Geography in the Java geometry libraries, but Hibernate Spatial (mostly) works with both. So in order not to confuse users I think on balance it is justified to have special functions for these specific postgis operators.

**Reply by beikov** [reply]
Now that we have [HHH-15160](https://hibernate.atlassian.net/browse/HHH-15160) for tracking this, I'll close this discussion.
